sklearn/tree/meson.build

```diff
@@ -0,0 +1,26 @@
+tree_extension_metadata = {
+  '_tree':
+    {'sources': ['_tree.pyx'],
+     'override_options': ['cython_language=cpp', 'optimization=3']},
+  '_splitter':
+    {'sources': ['_splitter.pyx'],
+     'override_options': ['optimization=3']},
+  '_criterion':
+    {'sources': ['_criterion.pyx'],
+     'override_options': ['optimization=3']},
+  '_utils':
+    {'sources': ['_utils.pyx'],
+     'override_options': ['optimization=3']},
+}
```

Comment:
we have a lot of boilerplate in the `meson.build` files, and I wish there was a way to have them all in the parent file, and in these files only include these dictionaries for the parent file to know what to do.

Reply:
> we have a lot of boilerplate in the meson.build files

We do have a lot of boiler-plate code indeed, suggestions more than welcome on how to improve this.

> I wish there was a way to have them all in the parent file, and in these files only include these dictionaries for the parent file to know what to do.

While working on this, I found the contrary, trying to read the single-file `setup.py` was a pain and I went so far as getting back the old distutils per sub-folder `setup.py` files (full disclosure: parts of this is distutils vs setuptools I think). This was a lot easier to have per-sub-package `meson.build` file that only cares about their sub-package.

I also found there are were plenty of tiny variations between sub-folders, optimization level, cython language, dependencies which are tricky to get right, generated pyx or not, generated pxd or not, etc ... Having the data in the sub-folders and the extension creation logic in the root `meson.build` file would not simplify the code at all.

Reply:
Came here because I am also curious if we can somehow reduce the number of `meson.build` files. Where is the code/config that currently configures everything? Is it the `extension_config` and `configure_extension_modules()` in `setup.py`?

Having a reference to how much "stuff" we currently have would help counter act my kneejerk reaction of "why the heck do I see meson meson meson everywhere now?!?" :D

In general I think having a build system/config that is simple but not easy to understand is probably good. Simple can mean a lot of repetition. The benefit would be that you only have to understand things locally to reason about a local thing. An easy system might do a lot of inference from some global config and "just work" because files and names are organised just right. The downside of this is that you have to understand the whole thing/global source of truth to reason about local stuff.

So I think having lots of repetition and boilerplate could be good (aka simple). But we have to make sure we create some source of knowledge for the little differences between the various `meson.build` files that, at first glance, all look the same (see Adrin's comments about "why is this like this here but not in other build files?"). Maybe we should have a set of comments for the (hopefully) finite number of "tricks" or deviations from the standard meson build file ("If you need this to be cimpart'able you need to add this and that")

Reply:
> Where is the code/config that currently configures everything? Is it the extension_config and configure_extension_modules() in setup.py?
> Having a reference to how much "stuff" we currently have would help counter act my kneejerk reaction of "why the heck do I see meson meson meson everywhere now?!?" :D

~400 lines of code in `setup.py` starting [here](https://github.com/scikit-learn/scikit-learn/blob/34061dafa67282e293b33d40f03b8ac60efeb2f1/setup.py#L193). Some of it is quite obscure i.e. I find the svm setup a lot more understandable in meson (and also in our previous distutils setup) than in setuptools.

It's funny because I come from the other direction, I looked at other projects like scikit-image which have a `meson.build` in every single folder including tests where they have listed explictly all the python files and I am like Phew I am glad I did not have to do this, see e.g.
https://github.com/scikit-image/scikit-image/blob/0e35d8040d0b0bb409de41f3c0f024f941e09d8a/skimage/color/meson.build#L4

By the way I have asked about the last point in the Scientific Python Discourse but nobody has bitten yet.

Reply:
You can avoid some boilerplate with `install_subdir`, which works okay for all pure Python directories. The tradeoff is that it's (just like with `setuptools`) basically globbing everything then, and it's then easy to get tmpfiles and other things you didn't mean to install into the installed package. So it's a tradeoff; both Meson and CMake recommend to avoid globbing and to explicitly list files - but it's a matter of pragmatism what you prefer.

When you do have compiled extensions, it's my experience that it's nicer to have them defined in a build config file in the same directory as they're going to be installed. There's a reason why the most popular feature of `numpy.distutils` was "nested `setup.py` files".

> By the way I have asked about the last point in the Scientific Python Discourse but nobody has bitten yet.

I usually don't follow along closely there, but I had a look and can't find your post - could you link to it?

Reply:
Thanks for your comments, this is super useful!

> I usually don't follow along closely there, but I had a look and can't find your post - could you link to it?

I think Stefan van der Walt has asked you about it and posted your reply since. I think for now I am happy to go the `install_subdir` route. For reference the Discord question was https://discord.com/channels/786703927705862175/1110263136176578592/1192785501862760478.

Reply:
> When you do have compiled extensions, it's my experience that it's nicer to have them defined in a build config file in the same directory as they're going to be installed. There's a reason why the most popular feature of `numpy.distutils` was "nested `setup.py` files".

Fully agree on this!

Reply:
You originally said "Discourse", now it is "Discord". :)

Reply:
Ah yes, I remember Stefan asking about that. I'm not on Discord - can only do so many real-time communication channels:)